[영화 극장 선택] 팡태 3, 4단계 미션 제출합니다

Hevton opened the pull request:
안녕하세요 제이든! 3,4 단계 미션도 잘 부탁드립니다 😄😄 !!


https://github.com/woowacourse/android-movie-theater/assets/86276828/887afabf-3e5b-4b9f-8b61-a8c840b92aaf

murjune committed:
[영화 극장 선택] 오둥이 1, 2단계 미션 제출합니다. (#62)

* :tada: Theater시작!

* :sparkles: 영화 선택 후 극장을 선택할 수 있음,
극장 이름 줄이기

* :recycle: 분기처리

* :recycle: dataBinding 적용

* :truck: seat->activity_theater_seat

* :truck: seat->activity_theater_seat

* :recycle: Test오류 고침

* :sparkles: 영화 시간 정보 최종 화면에 표시

* :bug: 카운팅 안되던 버그 고침

* build: resources 파일 exclude 할 목록 추가

* ui: 바텀 네비 menu

* ui: 예약, 영화 목록, 세팅 Fragment

* ui: Fragment transaction 적용

* feat: MovieListFragment 로 마이그레이션

* :recycle:  findFragmentByTag 로 개선

* :bug: ticket count 로직 개선

* :memo: 기능 요구사항 명세서 업데이트

* :white_check_mark: MovieListFragment test

* :truck: 이름 변경

* :truck: xml 이름 통일성

* :sparkles: timeSpinner 개선

* :recycle: TheaterViewHolder dataBinding 적용

* :recycle: MovieDetailActivity dataBinding으로 개선

* :recyle: seat activity databinding apply

* :recyle: Detail layout Spinner 위치 조정

* :recycle: :white_check_mark: PurchaseConfirmationActivityTest 구조 변경

- 액티비티 변경에 따른 테스트 코드 변경

* :recycle: :white_check_mark: MovieListFragmentTest 구조 변경

* :recycle: :white_check_box: 테스트 코드 고침

* :recycle: ktLint Check

* :truck: 테스트 코드 패키지 프로덕트와 동일하게 맞춤

* :memo: README.md 할 일 전부 체크 완료

* :recycle: 프레젠터나 뷰가 값을 반환하지 않음

* :recycle: 네임스페이스는 모든 태그의 최상단에 작성

* :recycle: MovieDetailContract.kt 반환값 모두 삭제

* :recycle: onViewCreated() 다이어트

* :recycle: ktFormating

* :white_check_box: :recycle: 화면 이동 테스트 IntegrationTest.kt로 이동

* :sparkles: Intent오류 발생시 오류 Activity로 이동

* :sparkles: Activity, Fragment 바인딩기능 추가

* :recycle: Activity, Fragment 바인딩기능 적용

* :recycle: Cinema 프레젠터 내부로 이동

* :recycle: onClicked renameing

* :sparkles: Fragment redirectToErrorActivity

* :sparkles: TheatersBottomSheet MVP 구조 적용

* :recycle: MovieListPresenter.kt
loadMovie동작 프레젠터 내에서 init으로 적용

* :truck: showBottomSheet->navigateToCinemaView

* :truck: generateDateRange->loadDateRange

* :recycle: ktFormat

* :bug: init으로 옯긴 후 private val 위치 변경

* :recycle: 어댑터 구조 변경

* :recycle: 인텐트 전달 방식 변경

* :recycle: SeatPresenterTest capture 도입

* :recycle: fragment Container 이름 변경

* :recycle: timeDate formatting 개선

* :recycle: Time/Date Adapter 개선

* :recycle: ktlintCheck

* :truck: BindingActivity base package로 이동

* :recycle: ErrorActivity 시작 로직으로 변경

* :bug: Time/Date items MutableList로 변경

---------

Co-authored-by: 박재성 <jason@woowahan.com>
Co-authored-by: ii2001 <iivks2001@gmail.com>
Co-authored-by: hevton <hevton22@gmail.com>

Hevton committed:
[영화 극장 선택] 팡태 1, 2단계 미션 제출합니다 (#66)

murjune committed:
test: Mockk study

murjune committed:
test: Mockk study

murjune committed:
:truck: package 구조 정리

murjune committed:
:truck: package 구조 정리

murjune committed:
docs: 3,4 단계 기능명세서 작성

murjune committed:
docs: 3,4 단계 기능명세서 작성

murjune committed:
:recycle: item_movie_list ui 수정

murjune committed:
:recycle: item_movie_list ui 수정

murjune committed:
🐛  MovieList package fix

murjune committed:
🐛  MovieList package fix

murjune committed:
✨  Movie DB 만들기

  - [x]: Movie DB 만들기
  - [x]: Reservation Table(=Entity) 만들기
  - [x]: Seat Table(=Entity) 만들기
  - [x]: dao
  - [x]: dao Test

murjune committed:
✨  Movie DB 만들기

  - [x]: Movie DB 만들기
  - [x]: Reservation Table(=Entity) 만들기
  - [x]: Seat Table(=Entity) 만들기
  - [x]: dao
  - [x]: dao Test

murjune committed:
✨  MovieMapper, Reservation model 생성

murjune committed:
✨  MovieMapper, Reservation model 생성

murjune committed:
✨  MovieRepository

murjune committed:
✨  MovieRepository

murjune committed:
✨  TheaterSeatActivity DB 연결

- [x]: Mapper (Entity -> Domain Model)
 - [x]: Movie Repository
- [x]: 좌석 선택 뷰에서 예매 확인 시 Room 에 저장 하고 return 받은 id를 넘겨 주자
  - [x]: 저장은 worker Thread, startActivity 는 ui Thread 에서 하자고~
 - [x]: 기존 객체 직렬화 방식 -> id 넘겨 주는 방식으로 변경

murjune committed:
✨  TheaterSeatActivity DB 연결

- [x]: Mapper (Entity -> Domain Model)
 - [x]: Movie Repository
- [x]: 좌석 선택 뷰에서 예매 확인 시 Room 에 저장 하고 return 받은 id를 넘겨 주자
  - [x]: 저장은 worker Thread, startActivity 는 ui Thread 에서 하자고~
 - [x]: 기존 객체 직렬화 방식 -> id 넘겨 주는 방식으로 변경

murjune committed:
✨  예매 내역 리사이클러 뷰 만들기

  - [x]: Adapter
  - [x]: ItemView xml
  - [x]: ViewHolder

murjune committed:
✨  예매 내역 리사이클러 뷰 만들기

  - [x]: Adapter
  - [x]: ItemView xml
  - [x]: ViewHolder

murjune committed:
✨ NotificationDataStore

-  [x] SharedPref 에 푸시 알림 여부를 저장 한다.

murjune committed:
✨ NotificationDataStore

-  [x] SharedPref 에 푸시 알림 여부를 저장 한다.

murjune committed:
임시 저장

murjune committed:
:sparkles: postNotification 구현

murjune committed:
✨permission 분기 처리

- 푸시 알림 퍼미션 로직
-
    1) Home - 맨 처음에만 퍼미션 체크를 함
-
    2) Setting

    - 1번째 요청 아무 것도 없음
    - 2번째 요청에만 Explanation Dialog 보여줌
    - 3번째 요청부터도 아무 것도 없음 (이게.. 1, 3 번째 구분하는게 모르겠음)

murjune committed:
✨permission 분기 처리

- 푸시 알림 퍼미션 로직
-
    1) Home - 맨 처음에만 퍼미션 체크를 함
-
    2) Setting

    - 1번째 요청 아무 것도 없음
    - 2번째 요청에만 Explanation Dialog 보여줌
    - 3번째 요청부터도 아무 것도 없음 (이게.. 1, 3 번째 구분하는게 모르겠음)

Commit:
:white_check_mark: Presenter 테스트 구현

Commit:
:white_check_mark: Presenter 테스트 구현

murjune committed:
:truck: ReservationContract 패키지 분리

murjune committed:
:truck: ReservationContract 패키지 분리

murjune committed:
:recycle: reservation id 수정

- rv_movies -> rv_reservation_movies

murjune committed:
:recycle: reservation id 수정

- rv_movies -> rv_reservation_movies

murjune committed:
:sparkles: MovieDetailActivityTest

murjune committed:
:sparkles: MovieDetailActivityTest

murjune committed:
:recycle: 함수 분리

murjune committed:
:recycle: 함수 분리

murjune committed:
:recycle: 30분 전 푸시 알림, custom icon 적용

murjune committed:
:recycle: 30분 전 푸시 알림, custom icon 적용

murjune committed:
:bug: MovieListFragmentTest

- 멀티 뷰타입으로 바뀐 다음 테스트 깨지는거 수정

murjune committed:
:bug: MovieListFragmentTest

- 멀티 뷰타입으로 바뀐 다음 테스트 깨지는거 수정

murjune committed:
:sparkles: 32 일 때는 preference canNotify 를 true, 33 이상은 false 로 변경

murjune committed:
:sparkles: 32 일 때는 preference canNotify 를 true, 33 이상은 false 로 변경

murjune committed:
:truck: PurchaseConfirmation packaging

murjune committed:
:truck: PurchaseConfirmation packaging

murjune committed:
:sparkles: MovieRepositoryStore 구현

- 동적으로 MovieRepository를 주입하기 위해 구현

murjune committed:
:sparkles: MovieRepositoryStore 구현

- 동적으로 MovieRepository를 주입하기 위해 구현

murjune committed:
:sparkles: PurchaseConfirmationActivityTest 개선

murjune committed:
:sparkles: PurchaseConfirmationActivityTest 개선

murjune committed:
:recycle: ktLintFormat

murjune committed:
:recycle: ktLintFormat

murjune committed:
docs: README 수정

질문 사항

murjune committed:
docs: README 수정

질문 사항

murjune committed:
:recycle: 리뷰 반영

murjune committed:
:recycle: 리뷰 반영

KwonDae commented on app/src/main/java/woowacourse/movie/presentation/setting/SettingFragment.kt:
```diff
@@ -0,0 +1,84 @@
+                notificationPreference.canNotification = true
+                binding.switchAlarm.isChecked = true
+            } else {
+                showToast(getString(R.string.deny_notification_permission))
```
권한 flow 별로 구현을 정말 잘 해주셨네요 👍 

현재 권한이 두번의 거절상태이고, 그럼에도 유저가 활성화를 원한다면 앱 설정 화면으로 보내는것도 현업에서 자주 사용하는 UX인데요,

더이상 권한 요청을 하지 못하는 상황에서도 활성화를 시키는 유저를 어떻게 처리할지는
[ACTION_APPLICATION_DETAILS_SETTINGS](https://developer.android.com/reference/android/provider/Settings.html#ACTION_APPLICATION_DETAILS_SETTINGS) 키워드로 찾아보시면 알 수 있을거에요.

이부분은 어려운 내용일 수 있으니 고민되는 부분이 있다면 언제든 DM 주세요!

Hevton replied:
감사합니다 제이든!! 이 부분 그렇지 않아도 저희가 엄청 애를 먹었는데요 ㅜㅜ
유저에게 처음 권한을 요청할 때의 상황(isGranted: false, shouldShowRequestPermissionRationale: false)과 
두 번 이상 거절했을 때의 상황(isGranted: false, shouldShowRequestPermissionRationale: false)을 구분하기가 어렵다고 판단하여 SharedPreference를 추가로 이용해야만 구현이 가능하다고 생각하여 그렇게 구현했습니다!! 혹시 이 모든 분기문을 추가적인 데이터 저장 없이 진행할 수 있는 방안이 있는지 여쭤볼 수 있을까요?

https://github.com/woowacourse/android-movie-theater/assets/86276828/bdb6ba6c-163f-476f-a5e8-3a3cfc1b30bf

KwonDae replied:
그래서 보통 권한이 필요한 해당 화면에 진입했을 때 유저가 처음 권한 팝업을 맞이하도록 한번 띄워줍니다.
그래야 토글로 사용자가 다시 ON 상태를 원할 때 권한 체크 케이스를 좁힐 수 있거든요.
유저가 토글을 건드렸다는거 자체가 권한을 처음 마주하는 케이스가 아니기에,
그때는 isGranted 와 shouldShowRequestPermissionRationale 조합으로 유저가 두번이상 거절했다는걸 알 수 있습니다.

KwonDae commented on app/src/androidTest/java/woowacourse/movie/data/dao/MovieDaoTest.kt:
```diff
@@ -0,0 +1,83 @@
+    }
+
+    @Test
+    @DisplayName("예매 내역을 저장 하고, id로 영화 내역을 조회할 수 있다.")
+    fun insert_and_load_test() {
+        repeat(1000) {
+            // given
+            val reservationEntity = reservationEntity().copy(cinemaName = "$it")
+            // when
+            val id: Long = movieDao.saveReservation(reservationEntity)
+            val reservationResult = movieDao.loadReservation(id)
+            // then
+            assertSoftly {
+                id shouldBe (it + 1L)
+                reservationResult shouldBe reservationEntity
+            }
+        }
+    }
```
DB테스트도 id를 기반으로 저장을 하고 조회 쿼리를 통해 테스트 좋습니다 👍

KwonDae commented on app/src/main/java/woowacourse/movie/MovieReservationApp.kt:
```diff
@@ -0,0 +1,36 @@
+import woowacourse.movie.data.datastore.DefaultNotificationDataStore
+
+class MovieReservationApp : Application() {
+    val movieRepository: MovieRepository by lazy {
+        MovieRepositoryStore.instance(applicationContext).movieRepository()
+    }
```
```kotlin
val movieRepository: MovieRepository by lazy {
        DefaultMovieRepository(MovieDatabase.instance(this).movieDao())
    }
```

위와 같이 가능할 것 같은데 `MovieRepositoryStore`의 역할이 무엇인가요?

Hevton replied:
App에서 사용하는 Repository를 Ui Test에서 테스트용 Repository로 동적으로 변경해주기 위해서 이러한 방식을 적용했습니다!

KwonDae commented on app/src/main/java/woowacourse/movie/data/DefaultMovieRepository.kt:
```diff
@@ -0,0 +1,31 @@
+import woowacourse.movie.model.Reservation
+
+class DefaultMovieRepository(
+    private val movieDao: MovieDao,
```
DB를 repository에서 접근하는 방식 좋습니다 👍

KwonDae commented on app/src/main/java/woowacourse/movie/data/MovieRepositoryStore.kt:
```diff
@@ -0,0 +1,46 @@
+import woowacourse.movie.data.dao.MovieDao
+import woowacourse.movie.data.db.MovieDatabase
+
+class MovieRepositoryStore(private val context: Context) {
```
위에서도 코멘트를 드렸었는데요,
이 클래스의 역할이 어떤건지 잘 모르겠습니다 🥲

KwonDae commented on app/src/main/java/woowacourse/movie/data/dao/MovieDao.kt:
```diff
@@ -0,0 +1,47 @@
+    @Insert(onConflict = OnConflictStrategy.IGNORE)
+    fun saveMovieSeats(seats: List<SeatEntity>)
+
+    @Query(
+        """
+        SELECT * FROM 
+        (SELECT * FROM reservation WHERE id = :id) as res
+        JOIN movie_seats AS seat
+        ON res.id = seat.reservationId
+        ORDER BY seat.rowNumber, seat.columnNumber
+    """,
+    )
+    fun loadReservationWithSeats(id: Long): Map<ReservationEntity, List<SeatEntity>>
+
```
id를 쿼리문을 통해 조회하도록 작성해주신게 인상깊었네요

KwonDae commented on app/src/main/java/woowacourse/movie/data/entity/ReservationEntity.kt:
```diff
@@ -0,0 +1,16 @@
+    val runningTime: String,
+    val releaseDate: String,
+) {
+    @PrimaryKey(autoGenerate = true)
+    var id: Long = 0
```
PrimaryKey인 id를 수정가능하도록 선언하신 이유가 있을까요?
https://developer.android.com/training/data-storage/room/defining-data?hl=ko#anatomy

Hevton replied:
찾아본 바로는 auto generate 설정이 적용되려면 이렇게 해야 하는 것으로 이해했는데, 다른 방법이 있을까요?

KwonDae commented on app/src/main/java/woowacourse/movie/presentation/purchaseConfirmation/PurchaseConfirmationActivity.kt:
```diff
@@ -0,0 +1,68 @@
+import woowacourse.movie.presentation.error.ErrorActivity
+import kotlin.concurrent.thread
+
+class PurchaseConfirmationActivity :
```
현재 예매완료 페이지에서 뒤로가기를 하면 좌석선택 화면으로 돌아가는데요,
사용자는 홈으로 가기를 기대하고 있지 않을까요?

Hevton replied:
감사합니다 이를 적용해보겠습니다!

KwonDae commented on app/src/main/java/woowacourse/movie/presentation/reservation/ReservationAdapter.kt:
```diff
@@ -0,0 +1,54 @@
+    }
+
+    @SuppressLint("NotifyDataSetChanged")
+    fun updateItems(reservations: List<Reservation>) {
+        items = reservations
+        notifyDataSetChanged()
+    }
```
👍

KwonDae commented on app/src/main/java/woowacourse/movie/presentation/setting/SettingFragment.kt:
```diff
@@ -0,0 +1,84 @@
+        binding.switchAlarm.isChecked = isChecked
+    }
+
+    private fun isSecondRequestPermission(): Boolean =
+        (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) &&
+            shouldShowRequestPermissionRationale(
+                POST_NOTIFICATIONS,
+            )
```
`shouldShowRequestPermissionRationale` 함수는 사실 사용자가 권한을 한번 거부했고, 이후에 다시 거부하거나 허용을 할 때 까지 계속 true로 반환되는 함수입니다.

예를 들어, 권한 팝업 화면에서 뒤로가기를 눌러 팝업을 닫아버리면, 다음에 해당 화면 진입시에 `shouldShowRequestPermissionRationale` 는 true를 반환합니다. 
https://developer.android.com/training/permissions/requesting?hl=ko#explain

Hevton replied:
감사합니다 이를 활용했습니다!

KwonDae commented on app/src/main/java/woowacourse/movie/MovieBroadCastReceiver.kt:
```diff
@@ -0,0 +1,80 @@
+        }
+    }
+
+    private fun sendNotification(
+        context: Context,
+        movieTitle: String,
+        reservationId: Long,
+    ) {
+        createNotificationChannel(context)
+        val pendingIntent = createPendingIntent(context, reservationId)
+
+        val notificationBuilder =
+            NotificationCompat.Builder(context, RESERVATION_NOTIFICATION_ID)
+                .setContentTitle("예매 알림")
+                .setContentText("$movieTitle 30분 후에 상영")
+                .setSmallIcon(R.drawable.ic_notifications_active_24)
+                .setContentIntent(pendingIntent)
+                .setPriority(NotificationCompat.PRIORITY_DEFAULT)
+                .setAutoCancel(true)
+
+        val notificationManager =
+            ContextCompat.getSystemService(context, NotificationManager::class.java)
+        notificationManager?.notify(RESERVATION_REQUEST_CODE, notificationBuilder.build())
+    }
```
`createNotificationChannel` 함수를 분리해주신 것 처럼
각 기능별로 함수를 분리해보면 가독성이 좋아지겠네요 👍

KwonDae commented on app/src/main/java/woowacourse/movie/data/datastore/DefaultNotificationDataStore.kt:
```diff
@@ -0,0 +1,35 @@
+
+    private fun defaultCanNotification(): Boolean = Build.VERSION.SDK_INT < Build.VERSION_CODES.TIRAMISU
+
+    companion object {
+        @Volatile
+        private var instance: NotificationDataStore? = null
+        private const val DATASTORE_KEY = "DATASTORE_KEY"
+        private const val NOTIFICATION_KEY = "NOTIFICATION_KEY"
+
+        fun instance(context: Context): NotificationDataStore {
+            return instance ?: synchronized(this) {
+                instance ?: DefaultNotificationDataStore(
+                    context.getSharedPreferences(DATASTORE_KEY, Context.MODE_PRIVATE),
+                ).also { instance = it }
+            }
+        }
+    }
```
현재 DefaultNotificationDataStore는 Application에서 생성하여 전역에서 접근하여 사용하고 있습니다.
Application에서 생성하여 전역에서 접근하기에 단일 인스턴스를 사용하게 되는데요, 중복으로 singleton 처리를 해준 이유가 있을까요?

Hevton replied:
감사합니다. 중복으로 singletone 처리를 진행했네요..! application 단에서 생성하지 않고, 필요한 곳에서만 싱글톤으로 사용할 수 있게끔 진행했습니다!

KwonDae commented on app/src/main/java/woowacourse/movie/data/MovieRepositoryStore.kt:
```diff
@@ -0,0 +1,46 @@
+        movieDao = null
+    }
+
+    companion object {
+        @Volatile
+        private var instance: MovieRepositoryStore? = null
+
+        fun instance(context: Context): MovieRepositoryStore {
+            return instance ?: synchronized(this) {
+                instance ?: MovieRepositoryStore(context).also { instance = it }
+            }
+        }
+    }
```
DefaultNotificationDataStore 코멘트와 마찬가지입니다!

KwonDae reviewed (changes requested):
3,4단계를 정말 깔끔하게 잘 구현해주셨네요 👍 
전체적으로 깔끔하게 분리된 구조가 인상깊었습니다.

논의하고 싶은 부분에 코멘트를 남겨드렸으니
확인 후 요청 부탁드립니다!

murjune committed:
:recycle: NotificationDataStore 프로퍼티 네이밍 변경

canNotification -> acceptedPushAlarm

murjune committed:
:sparkles: SettingFragment 완전히 거절 시 App Setting으로 이동

murjune committed:
:recycle: MovieDetailContract 네이밍 변경

- showReleaseDates
-  showReleaseTimes

murjune committed:
:truck: testApplicationContext 으로 네이밍 변경

murjune committed:
:fire: MovieDaoTest repeat(1000) 삭제

murjune committed:
:sparkles: 예매완료 페이지에서 뒤로가기 시 Home 으로 이동

murjune committed:
:recycle: MovieBroadCastReceiver 함수 분리

murjune committed:
:recycle: MovieRepositoryInjector 로 리팩토링

- 중복해서 싱글톤으로 만드는 로직 개선(App 에 있던 로직 삭제)
- MovieRepositoryStore ---> MovieRepositoryInjector 로 네이밍 변경

murjune committed:
:recycle: ktLintFormat

Commit:
Merge branch 'step3' of https://github.com/Hevton/android-movie-theater into step3

# Conflicts:
#	README.md
#	app/src/androidTest/java/woowacourse/movie/data/dao/MovieDaoTest.kt
#	app/src/androidTest/java/woowacourse/movie/presentation/movieDetail/MovieDetailActivityTest.kt
#	app/src/androidTest/java/woowacourse/movie/presentation/purchaseConfirmation/PurchaseConfirmationActivityTest.kt
#	app/src/androidTest/java/woowacourse/movie/util/ContextUtils.kt
#	app/src/main/AndroidManifest.xml
#	app/src/main/java/woowacourse/movie/MovieBroadCastReceiver.kt
#	app/src/main/java/woowacourse/movie/MovieReservationApp.kt
#	app/src/main/java/woowacourse/movie/data/datastore/DefaultNotificationDataStore.kt
#	app/src/main/java/woowacourse/movie/data/datastore/NotificationDataStore.kt
#	app/src/main/java/woowacourse/movie/presentation/HomeActivity.kt
#	app/src/main/java/woowacourse/movie/presentation/movieDetail/MovieDetailActivity.kt
#	app/src/main/java/woowacourse/movie/presentation/movieDetail/MovieDetailContract.kt
#	app/src/main/java/woowacourse/movie/presentation/movieDetail/MovieDetailPresenter.kt
#	app/src/main/java/woowacourse/movie/presentation/purchaseConfirmation/PurchaseConfirmationActivity.kt
#	app/src/main/java/woowacourse/movie/presentation/reservation/ReservationFragment.kt
#	app/src/main/java/woowacourse/movie/presentation/seat/TheaterSeatActivity.kt
#	app/src/main/java/woowacourse/movie/presentation/setting/SettingFragment.kt
#	app/src/main/res/values/strings.xml
#	app/src/test/java/woowacourse/movie/presentation/movieDetail/MovieDetailPresenterTest.kt

KwonDae commented on app/src/main/java/woowacourse/movie/data/datastore/DefaultNotificationDataStore.kt:
```diff
@@ -0,0 +1,44 @@
+        private const val NOTIFICATION_KEY = "NOTIFICATION_KEY"
+        private const val POST_NOTIFICATIONS = "POST_NOTIFICATIONS"
+
+        fun instance(context: Context): NotificationDataStore {
+            return instance ?: synchronized(this) {
+                instance ?: DefaultNotificationDataStore(
+                    context.getSharedPreferences(DATASTORE_KEY, Context.MODE_PRIVATE),
+                ).also { instance = it }
+            }
+        }
+    }
```
https://github.com/woowacourse/android-movie-theater/pull/86#discussion_r1597601742

사실 이 코멘트는 Application에서 처리하면서 application context 와 activity context 에 생각을 어떻게 하셨을까 궁금하여
 이야기를 나누고 싶었던 코멘트 였습니다 🥲

Sharedpreferences, DataStore, Room 은 둘 중 어느 context를 필요로 할까요? [참고글](https://stackoverflow.com/questions/11567134/sharedpreferences-application-context-vs-activity-context)
또한, Repository 와 DataStore는 어떤 context를 필요로 할까요?

지금은 대부분 Activity내에서 applicationContext를 가져오고 있는데요
activity외의 영역에서 context가 필요한 경우 application context를 사용하도록 실제로 [가이드](https://developer.android.com/codelabs/android-room-with-a-view-kotlin?_gl=1*123trwt*_up*MQ..*_ga*NDc3MDYwMjQ1LjE3MTU1NzQ0MDc.*_ga_6HH9YJMN9M*MTcxNTU3NDQwNy4xLjAuMTcxNTU3NDQwNy4wLjAuMA..#12)하고 있습니다.

저 또한 팡태가 이전에 하셨던 방법인 application 클래스에서 주입을 해주는 방법은 좋다고 생각합니다. 실제로 다수 라이브러리에서 context가 필요한 경우 application에서 설정을 요구하곤 합니다.

위 내용에 대해 고민 해보시고 반영해주시면서 생각을 공유해주세요!

Hevton replied:
안녕하세요 제이든, 남겨주신 코멘트 복기하였습니다.
아직도 application context와 activity context에 대한 확신이 부족한 것 같습니다만
애플리케이션의 전체 라이프사이클 동안 유지되어야 하는 데이터나 리소스에 접근할 때는 applicationContext를, 
액티비티와 관련된 리소스에 접근할 때는 activity context를 사용해야겠다는 직감이 있습니다.

Application 단에서 Activity의 context를 사용하는 것은, 굉장히 위험한 일이라고 생각합니다 
(App의 생명주기가 Activity보다 길기 때문에 이미 죽어버린 액티비티를 참조할 수 있기 때문에)

하지만 Activity에서, 언제 application context를 사용하고 언제 activty context를 사용할지에 대해서는 아직 뚜렷한 확신이 서지는 않는데요.. 
Application 에서 applicationContext로 Repository를 생성한다면 앱 전역동안 살아있어서 어느 액티비티에서라도 가져다 쓸 수 있을 것 같다는 생각은 들지만,
만약 Activity에서도 앱 전역동안 살아있는 싱글톤 객체를 만들고 싶을 때엔 applicationContext를 참조해서 생성하면 되는 것일까요?

KwonDae replied:
> 만약 Activity에서도 앱 전역동안 살아있는 싱글톤 객체를 만들고 싶을 때엔 applicationContext를 참조해서 생성하면 되는 것일까요?

그렇다면 Application에서 싱글톤 객체를 생성해주면 해결되지 않을까요~?
아직 활용을 많이 해보지 않으셔서 Application context와 Activity context에 대한 감이 잘 안오실 수 있어요.
사실 Application에서 Activity context를 참조할 수 있을까? 생각해보면 Application에서는 앱이 가장 처음에 구동될 때 Activity보다 먼저 생성되다보니 Activity context에 접근할 일이 없긴 합니다.

Application에 대한 [공식문서](https://developer.android.com/reference/android/app/Application)를 읽어보셔도 도움이 될거에요

KwonDae reviewed (changes requested):
빠르게 피드백 반영을 잘 해주셨네요 👍 
마지막 요청이 될 것 같은데요,
다음 미션으로 넘어가기 전에 논의하고 넘어가야 할 부분이 있어서 해당 부분에 코멘트를 남겨드렸습니다.
확인 후 요청 부탁드립니다!

KwonDae reviewed (approved):
많은 피드백을 요청드렸음에도 잘 해내주셔서 감사합니다.
끝까지 남겨드린 피드백을 복기하는 과정들이 팡태에게 도움이 되었으면 좋겠네요.
영화 극장 미션 진행하시느라 고생많으셨고, 남은 미션들도 파이팅입니다!
관련해서 궁금한 부분은 언제든 편하게 DM주세요!
